### PR TITLE
Move torch/autograd/grad_mode.pyi stubs inline

### PR DESCRIPTION
**Pull request by rgommers**
- Add `torch._C` bindings from `torch/csrc/autograd/init.cpp`
- Renamed `torch._C.set_grad_enabled` to `torch._C._set_grad_enabled`
  so it doesn't conflict with torch.set_grad_enabled anymore

This is a continuation of gh-38201. All I did was resolve merge conflicts and finish the annotation of `_DecoratorContextManager.__call__` that @ezyang started in the first commit.

~Reverts commit b5cd3a80bbc, which was only motivated by not having `typing_extensions` available.~ (JIT can't be made to understand `Literal[False]`, so keep as is).

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by SplitInfinity** [comment]
I am waiting for a final review of my JIT-side changes on #41371 before I can land it.

**Comment by SplitInfinity** [comment]
#41371 has landed. Feel free to rebase and merge.

**Pull request closed without merging**
